Add data race detector plugin

AlphaDaze opened the pull request:
This PR adds a data race detector plugin. This plugin detects possible data races within the underlying program along with information that can be utilised for debugging.

The README provides further context for the plugin and its usage.

Compile with `make datarace_ft` and run with `mambo_datarace_ft`. The plugin includes two algorithms which can be compiled separately: `fasttrack` and `djit`. Both are happens-before algorithms with `fasttrack` being an improvement over `djit` with both memory and runtime.

All comments and suggestions are welcome!

**Sample output**

```
--- FORK: Thread 1029188 created by 0
--- LOCK: Thread 1029188 locked 0x7fff22128b10
--- UNLOCK: Thread 1029188 unlocked 0x7fff22128b10
--- FORK: Thread 1029189 created by 1029188
--- LOCK: Thread 1029188 locked 0x7fff22128b10
--- UNLOCK: Thread 1029188 unlocked 0x7fff22128b10
--- FORK: Thread 1029223 created by 1029188
--- JOIN: Thread 1029189 joining 1029188
!!! READ RACE: Possible read race - 0x41104c by thread 1029223
>>>> at 0x40074c: /home/user/Documents/git/tests/hg04_race (0x400744) (th10)
!!! WRITE RACE: Possible write race - 0x41104c by thread 1029223
>>>> at 0x40075c: /home/user/Documents/git/tests/hg04_race (0x400744) (th10)
--- JOIN: Thread 1029223 joining 1029188
--- LOCK: Thread 1029188 locked 0x7fff22128ab0
--- UNLOCK: Thread 1029188 unlocked 0x7fff22128ab0
We're done; exiting with status: 0
```

IgWod reviewed (changes requested):
Hi @AlphaDaze,

Thank you for your contribution. It looks good, however there are few things I'd like to be addressed before merging it. Please see attached comments.

Also, a major point, please remove any commented out code or add explanation why it's commend out and how it can be used in the future. We don't need any dead code in the repo :).

Let me know if something is unclear.

Many thanks,
Igor

AlphaDaze commented on dbm.h:
```diff
@@ -274,6 +274,9 @@ struct dbm_thread_s {
   void *pstack;
+
+  // Can be utilised to trasnfer data to child thread
+  void *shared_plugin_data;
```
I agree. Renamed to `shared_parent_data`.

AlphaDaze commented on plugins/datarace/README.md:
```diff
@@ -0,0 +1,49 @@
+
+    ./mambo_datarace_ft /usr/bin/lscpu
+
+> **Note:** Underlying programs must be compiled with `-no-pie` (statically or dynamically) OR must be dynamically compiled and mambo compiled with `NO_PIE_ENABLE` disabled in [datarace.h](datarace.h)
```
Whoops... Fixed

AlphaDaze commented on plugins/datarace/README.md:
```diff
@@ -0,0 +1,49 @@
+> **Note:** Underlying programs must be compiled with `-no-pie` (statically or dynamically) OR must be dynamically compiled and mambo compiled with `NO_PIE_ENABLE` disabled in [datarace.h](datarace.h)
+
+
+Test applications utilising multiple threads have been provided in the plugins/datarace/test/ directory. TODO
```
Removed

AlphaDaze commented on plugins/datarace/README.md:
```diff
@@ -0,0 +1,49 @@
+
+There are two separate algorithms to choose from for the datarace detector: [FastTrack](https://dl.acm.org/doi/abs/10.1145/1543135.1542490) and [DJIT+](https://dl.acm.org/doi/abs/10.1145/781498.781529). Both algorithms utilise the happens-before relationship. FastTrack is known to be an improvement over DJIT+ implementations by carrying out a happens-before comparison in O(1) for the majority of  ...
+
+> To compile and use the DJIT+ implementation, use `make datarace_ft` and run with `./mambo_datarace_djit`.
```
Yup, fixed

AlphaDaze commented on plugins/datarace/README.md:
```diff
@@ -0,0 +1,49 @@
+
+Further algorithms can be implemented using the handling functins for the varoius thread operations at the beginnig of [datarace.c](datarace.c). Details of the implementation are also included at the top of the file for easier modification.
+
+[datarace.h](datarace.h) contains option to enable `no-pie` to properly ignore clib. More debug information can also be enabled.
```
Let's pretend I didn't make that mistake...multiple times 🙈

AlphaDaze commented on plugins/datarace/README.md:
```diff
@@ -0,0 +1,49 @@
+------------
+
+
+Currently clib is ignored by not processing any reads/writes to addresses above `0x7000000000` which require the underlying program to be compiled with `no-pie`. A workaround has been retrieve the location for dynamic libraries and ignore any files within /usr. There is likely a better solution to this however. Please don't hesitate to open a [pull request.](https://github.com/beehive-lab/mambo/p ...
```
Rephrased, let me know if it's still unclear please

AlphaDaze commented on plugins/datarace/datarace.c:
```diff
@@ -0,0 +1,431 @@
+  int tid = mambo_get_thread_id(ctx); // thread id
+
+  if (is_load || is_store) {
+    emit_push(ctx, (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3));
```
Changed throughout

AlphaDaze commented on plugins/datarace/datarace.c:
```diff
@@ -0,0 +1,431 @@
+  assert(set_cb == MAMBO_SUCCESS);
+  
+
+  set_cb = mambo_register_function_cb(ctx, "pthread_mutex_lock",
```
Changed indentation for function callbacks

AlphaDaze commented on plugins/datarace/datarace.h:
```diff
@@ -0,0 +1,38 @@
+
+
+typedef struct thread_data {
+  pthread_mutex_t *mutex; // change this to address val?
```
Removed comment

IgWod commented on plugins/datarace/datarace.c:
```diff
@@ -0,0 +1,435 @@
+}
+
+void process_read(int tid, uintptr_t ld_addr, uintptr_t inst_addr) {
+  // info("Processing: READ(%lx): %u %p\n", inst_addr, tid, ld_addr);
```
Why is it commented out?

AlphaDaze replied:
Uncommented and changed to debug output

IgWod commented on plugins/datarace/datarace.c:
```diff
@@ -0,0 +1,435 @@
+}
+
+void process_write(int tid, uintptr_t st_addr, uintptr_t inst_addr) {
+  // info("Processing: WRITE(%lx): %u %p\n", inst_addr, tid, st_addr);
```
Again, why is it commented out?

AlphaDaze replied:
Uncommented and changed to debug output

IgWod reviewed (changes requested):
Thanks for addressing the comments. I found two more small issues, but I'm happy otherwise. Please either remove the commented out code I highlighted or add TODO/explanation why it's commented out.

The other things is commit messages. Could you please clean it up? Ideally we would like to have two commits. The first commit for changes to the core and then the second one adding the plugin. Shouldn't be too difficult to achieve with `git rebase -i` and some amending.

Once that's done I'm happy with the PR, but I'll ask @jkressel to have a look as well, and once he approves it we can merge it.

AlphaDaze commented:
Sounds good! Thanks for taking a look at it. I have rebased and squashed the relevant commits together as requested.

If there is any other changes (however small), please do let me know.

josh-nook commented:
Looks good to me

IgWod commented:
Looks good to me! I'll ask @jkressel to have a look now.

jkressel reviewed (approved):
Looks good to me, I'm happy to merge.
